src/main/java/technology/dice/dicefairlink/discovery/members/FairlinkMemberFinder.java

```diff
@@ -0,0 +1,163 @@
+      return EMPTY_SET;
+    }
+  }
+
+  public final SizedIterator<String> discoverReplicas() {
+    try {
+      long before = System.currentTimeMillis();
+      final ClusterInfo clusterInfo = this.memberFinder.discoverCluster();
+      this.fallbackEndpoint = Optional.of(clusterInfo.getReadonlyEndpoint());
+      final Set<String> filteredReplicas =
+          clusterInfo.getReplicas().stream()
+              .filter(db -> !excludedInstanceIds.contains(db))
+              .filter(
+                  dbIdentifier ->
+                      (!this.fairlinkConfiguration.isValidateConnection())
+                          || this.validate(fairlinkConfiguration.hostname(dbIdentifier)))
+              .map(dbIdentifier -> fairlinkConfiguration.hostname(dbIdentifier))
+              .collect(Collectors.toSet());
+      final SizedIterator<String> result =
+          filteredReplicas.isEmpty()
+              ? this.iteratorBuilder.apply(this.setOf(clusterInfo.getReadonlyEndpoint()))
+              : this.iteratorBuilder.apply(filteredReplicas);
+      long after = System.currentTimeMillis();
+      LOGGER.info(
+          "Updated list of replicas in "
+              + (after - before)
+              + " ms. Found "
+              + filteredReplicas.size()
+              + " good, active, non-excluded replica"
+              + (filteredReplicas.size() != 1 ? "s" : "")
+              + " (validation "
+              + (fairlinkConfiguration.isValidateConnection() ? "" : "NOT ")
+              + "done). Excluded "
+              + this.excludedInstanceIds.size()
+              + " instance"
+              + (this.excludedInstanceIds.size() != 1 ? "" : "s")
+              + ". Next update in "
+              + this.fairlinkConfiguration.getReplicaPollInterval());
+      return result;
+    } catch (Exception e) {
+      LOGGER.log(
+          Level.WARNING,
+          "Error discovering cluster identified by ["
+              + this.fairlinkConnectionString.getFairlinkUri()
+              + "]. Will return fallback endpoint"
+              + this.fallbackEndpoint
+              + " if available",
+          e);
+      if (!this.fallbackEndpoint.isPresent()) {
+        LOGGER.log(
+            Level.SEVERE,
+            "Fallback endpoint not available. This means the cluster has never been successfully discovered. This is probably a permanent error condition");
+      }
+      return fallbackEndpoint
+          .map(fallbackEndpoint -> this.iteratorBuilder.apply(this.setOf(fallbackEndpoint)))
+          .orElseThrow(
+              () ->
+                  new RuntimeException(
+                      "Could not discover cluster identified by ["
+                          + fairlinkConnectionString.getFairlinkUri()
+                          + "] and a fallback reader endpoint is not available"));
+    }
+  }
+
+  private Set<String> setOf(String entry) {
+    Set<String> set = new HashSet<>(1);
+    set.add(entry);
+    return set;
```

Comment:
is this supposed to be mutable?

Reply:
This project doesn't have a dependency on Guava (only on test scope). Indeed it would be more elegant to use `ImmutableSet.of(entry)`

Reply:
but use `Collections.unmodifiable`